typed-racket-lib/typed-racket/types/pairwise-intersect.rkt

```diff
@@ -0,0 +1,121 @@
+#lang racket
+
+;; This module provides a function pairwise-intersect for calculating a lower
+;; bound of two types with respect to the precision order (*not* the subtype
+;; order). Contract typechecking uses this function to determine the type of a
+;; contracted value, given the value's type and the contract's output type.
+
+(require "../utils/utils.rkt"
+         (rep type-rep prop-rep object-rep values-rep)
+         (types subtype)
+         (only-in (infer infer) intersect)
+         racket/match)
+
+(provide pairwise-intersect)
+
+(define (pairwise-intersect/arr s t)
+  (match* (s t)
+    [((Arrow: s-dom s-rest s-kws s-rng)
+      (Arrow: t-dom t-rest t-kws t-rng))
+     #:when (= (length s-kws) (length t-kws))
+     (make-Arrow (map pairwise-intersect s-dom t-dom)
+                 (and s-rest t-rest
+                      (pairwise-intersect s-rest t-rest))
+                 (map pairwise-intersect/kw s-kws t-kws)
+                 (pairwise-intersect s-rng t-rng))]
+    [(_ _)
+     (raise-arguments-error
+      'pairwise-intersect/arr
+      "pairwise-interescting unsupported for drest/different length kws"
+      "s" s
+      "t" t)]))
+
+(define (pairwise-intersect/kw s t)
+  (match* (s t)
+    [((Keyword: ks ts rs) (Keyword: kt tt rt))
+     #:when (and (eq? kt ks)
+                 (or rt (not rs)))
+     (make-Keyword ks (pairwise-intersect ts tt) rs)]
+    [(_ _)
+     (raise-arguments-error
+      'pairwise-intersect/kw
+      "keywords must match"
+      "s" s
+      "t" t)]))
+
+(define (pairwise-intersect/prop-set ps qs)
+  (match* (ps qs)
+    [((PropSet: p+ p-) (PropSet: q+ q-))
+     (make-PropSet (pairwise-intersect/prop p+ q+)
+                   (pairwise-intersect/prop p- q-))]))
+
+(define (pairwise-intersect/prop p1 p2)
+  (match* (p1 p2)
+    [(p p) p]
+    [(p (TrueProp:)) p]
+    [((TrueProp:) p) p]
+    [((TypeProp: o s) (TypeProp: o t))
+     (make-TypeProp o (pairwise-intersect s t))]
+    [((NotTypeProp: o s) (NotTypeProp: o t))
+     (make-NotTypeProp o (pairwise-intersect s t))]
+    [(_ _)
+     (raise-arguments-error
+      'pairwise-intersect/prop
+      "cannot merge props"
+      "p1" p1
+      "p2" p2)]))
+
+;; pairwise-intersect : (case-> (-> Type Type Type)
+;;                              (-> Values Values Values)
+;;                              (-> Result Result Result))
+;; Computes a lower bound of the two types w.r.t. the "precision order." The
+;; precision order is like the subtype order except that it does not account for
+;; variance. Effectively, this amounts to a fold over the two types in a uniform
+;; way. For base types (and, as a default, for unimplemented cases) this
```

Comment:
Is there a brief thing we can insert here why we want to _ignore_ variance? Why it's sound? etc (it seems a little concerning to just read that we're ignoring variance w/o knowing why)

Reply:
I think the comment about ignoring variance shouldn't be there. I unfortunately don't have the theory worked out really well here, but I'll try to describe why I think the remark was misleading.

(First, maybe *pairwise* is the wrong modifier.. but bear with me) If we use **`p`** to refer to pairwise intersecting, then `(a -> b) `**`p`**` (c -> d)` is `(a `**`p`**` c) -> (b `**`p`**` d)`. I've been assuming **`p`** is commutative, so for the domains `a `**`p`**` c` should be the same as `c `**`p`**` a`, and so it seems like variance shouldn't even be brought up.